File: 2016/25_ClockSignal/assembunny.ts
// ======================================================================
// Clock Signal
//   Advent of Code 2016 Day 25 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           a s s e m b u n n y . t s
//
// A solver for the Advent of Code 2016 Day 25 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type RegisterName = 'a' | 'b' | 'c' | 'd' ;
type OpCode = 'cpy' | 'inc' | 'dec' | 'jnz' | 'out';
type Operand = RegisterName | number;
type Registers = Record<RegisterName, number>;
interface Instruction {
  opcode: OpCode;
  op1: Operand;
  op2?: Operand;
}
type Instructions = Instruction[];
type ProgramCounter = number;

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const registerLetters = 'abcd';

const nextOut = [1, 0];

// ======================================================================
//                                                             Assembunny
// ======================================================================

export class Assembunny {
  // Object for Clock Signal
  text: string[];

  part2: boolean;

  registers: Registers;

  instructions: Instructions;

  pc: ProgramCounter;

  output: number[];

  constructor(text: string[], part2 = false) {
    // Create a Assembunny object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.registers = {
      a: 0,
      b: 0,
      c: 0,
      d: 0,
    };
    this.output = [];
    this.instructions = [];
    this.pc = 0;

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText();
    }
  }

  processText(): void {
    // 1. Start with nothing
    this.instructions = [];
    // 2. Loop for each line of text
    this.text.forEach((line) => {
      // 3. Break the line into parts
      const [opcode, op1, op2] = line.split(' ');
      const xop1 = registerLetters.indexOf(op1) >= 0 ? op1 : parseInt(op1, 10);
      if (op2 && !op2.startsWith('#')) {
        const xop2 = registerLetters.indexOf(op2) >= 0 ? op2 : parseInt(op2, 10);
        this.instructions.push({
          opcode: <OpCode>opcode,
          op1: <Operand>xop1,
          op2: <Operand>xop2,
        });
      } else {
        this.instructions.push({
          opcode: <OpCode>opcode,
          op1: <Operand>xop1,
        });
      }
    });
  }

  resetRegisters(): void {
    this.registers = {
      a: 0,
      b: 0,
      c: 0,
      d: 0,
    };
    this.output = [];
    this.pc = 0;
  }

  currentState(): string {
    return `${this.pc}: ${this.currentRegisters()} ${this.currentInstruction()}`;
  }

  currentRegisters(): string {
    return `${this.registers.a} ${this.registers.b} ${this.registers.c} ${this.registers.d}`;
  }

  currentInstruction(): string {
    if (this.pc < 0 || this.pc >= this.instructions.length) return '';
    const inst = this.instructions[this.pc];
    switch (inst.opcode) {
      case 'cpy':
        return `cpy ${inst.op1} ${inst.op2}`;
      case 'inc':
      case 'dec':
      case 'out':
        return `${inst.opcode} ${inst.op1}`;
      case 'jnz':
        return `jnz ${inst.op1} ${inst.op2} (${<number>inst.op2 + this.pc})`;
      default:
        break;
    }
    return `unknown opcode: ${inst.opcode}`;
  }

  getOp1(inst: Instruction): number {
    return (typeof inst.op1 === 'number') ? inst.op1 : this.registers[<RegisterName>inst.op1];
  }

  step(verbose = false, stopOnOut = false): boolean {
    if (verbose) console.log(this.currentState());
    if (this.pc < 0 || this.pc >= this.instructions.length) return false;
    const inst = this.instructions[this.pc];
    const op1 = this.getOp1(inst);
    let nextPC = this.pc + 1;
    switch (inst.opcode) {
      case 'cpy':
        this.registers[<RegisterName>inst.op2] = op1;
        break;
      case 'inc':
        this.registers[<RegisterName>inst.op1] += 1;
        break;
      case 'dec':
        this.registers[<RegisterName>inst.op1] -= 1;
        break;
      case 'jnz':
        if (op1 !== 0) {
          nextPC = this.pc + <number>inst.op2;
        }
        break;
      case 'out':
        this.output.push(op1);
        if (stopOnOut) {
          return false;
        }
        break;
      default:
        console.log(`??? ${this.currentState()}`);
        return false;
    }
    this.pc = nextPC;
    return true;
  }

  run(verbose = false, limit = 0) {
    let steps = 0;
    while (this.step(verbose) && (limit === 0 || steps < limit)) {
      steps += 1;
    }
  }

  tryClockValue(a: number, verbose = false, limit = 0): boolean {
    if (verbose) console.log(`tryClockValue: a=${a} limit=${limit}`);
    const numBits = 10;
    let want = 0;
    let lastLength = 0;
    const maxSteps = limit > 0 ? limit : 9999;
    let steps = 0;
    this.resetRegisters();
    this.registers.a = a;
    while (this.output.length < numBits) {
      while (this.step(false, true) && steps < maxSteps);
      if (lastLength !== this.output.length - 1) {
        if (verbose) console.log(`tryClockValue: stopped at ${this.currentState()}`);
        return false;
      }
      if (this.output[lastLength] !== want) {
        if (verbose) console.log(`tryClockValue: wanted ${want} at ${this.currentState()}`);
        return false;
      }
      want = nextOut[want];
      lastLength += 1;
      this.pc += 1;
      steps += 1;
    }
    if (verbose) console.log(`tryClockValue: returning true at ${this.currentState()}`);
    return true;
  }

  findValueForClock(verbose = false, limit = 0) {
    if (verbose) console.log(`findValueForClock: ${limit}`);
    const maxA = limit === 0 ? 999 : limit;
    for (let a = 0; a < maxA; a += 1) {
      if (verbose) console.log(`findValueForClock a=${a}`);
      if (this.tryClockValue(a, verbose)) {
        return a;
      }
    }
    return NaN;
  }

  decoded(verbose = false, limit = 0) {
    if (verbose) console.log(`decoded: ${limit}`);
    const maxA = limit === 0 ? 200 : limit;
    for (let a = 0; a < maxA; a += 1) {
      const value = a + <number> this.instructions[1].op1 * <number> this.instructions[2].op1;
      const binary = value.toString(2).split('').reverse().join('');
      if (binary.length % 2 === 0 && binary[0] === '0') {
        if (verbose) console.log(`${a} ${value} ${binary}`);
        let need = '0';
        let ok = true;
        for (let index = 0; index < binary.length; index += 1) {
          if (binary[index] === need) {
            need = need === '0' ? '1' : '0';
          } else {
            ok = false;
          }
        }
        if (ok) {
          return a;
        }
      }
    }
    return NaN;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return this.decoded(verbose, limit);
    }
    return this.findValueForClock(verbose, limit);
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                    a s s e m b u n n y . t s                   end
// ======================================================================

File: 2016/25_ClockSignal/assembunny.py
# ======================================================================
# Clock Signal
#   Advent of Code 2016 Day 25 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         a s s e m b u n n y . p y
# ======================================================================
"A solver for the Advent of Code 2016 Day 25 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
TOGGLE = {
    'inc': 'dec',
    'dec': 'inc',
    'tgl': 'inc',
    'cpy': 'jnz',
    'jnz': 'cpy',
    'out': 'inc',
}
CLOCK_COUNT = 10
NEXT = [1, 0]
START_CLOCK = 0
LIMIT_CLOCK = 999

# ======================================================================
#                                                             Assembunny
# ======================================================================


class Assembunny(object):   # pylint: disable=R0902, R0205
    "Object for Safe Cracking"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.registers = {'a': 0, 'b': 0, 'c': 0, 'd': 0}
        self.program_counter = 0
        self.instructions = self.text
        self.output = []

    def reset(self):
        "Restart the computer"
        self.registers = {'a': 0, 'b': 0, 'c': 0, 'd': 0}
        self.program_counter = 0
        self.instructions = self.text
        self.output = []

    def fetch(self, arg):
        "Get the literal or register value"
        if arg.isalpha():
            return self.registers[arg]
        return int(arg)

    def step(self, verbose=False):  # pylint: disable=too-many-branches
        "Execute a single step of the computer"
        # 1. Can't execute instruction if pc too low or too high
        if self.program_counter < 0 or self.program_counter >= len(self.instructions):
            return False
        # 2. Fetch the instruction
        inst = self.instructions[self.program_counter].split()
        if verbose:
            print('pc: %d, %s' % (self.program_counter, ' '.join(inst)))
        # 3. Execute the instruction
        if inst[0] == 'cpy':
            if inst[2] in self.registers:
                self.registers[inst[2]] = self.fetch(inst[1])
            self.program_counter += 1
        elif inst[0] == 'inc':
            self.registers[inst[1]] += 1
            self.program_counter += 1
        elif inst[0] == 'dec':
            self.registers[inst[1]] -= 1
            self.program_counter += 1
        elif inst[0] == 'jnz':
            if self.fetch(inst[1]) != 0:
                self.program_counter += self.fetch(inst[2])
            else:
                self.program_counter += 1
        elif inst[0] == 'tgl':
            toggle_at = self.program_counter + self.fetch(inst[1])
            if 0 <= toggle_at < len(self.instructions):
                inst_old = self.instructions[toggle_at].split()
                inst_new = inst_old.copy()
                inst_new[0] = TOGGLE[inst_new[0]]
                self.instructions[toggle_at] = ' '.join(inst_new)
                if verbose:
                    print("Changed inst at %d from %s to %s" %
                          (toggle_at, ' '.join(inst_old), ' '.join(inst_new)))
            self.program_counter += 1
        elif inst[0] == 'out':
            self.output.append(self.fetch(inst[1]))
            self.program_counter += 1
        else:
            print("Unknown instruction at %d: %s" % (self.program_counter, inst))
            return False
        return True

    def run(self, verbose=False, initial_a=0):
        "Execute instructions until stopped"
        self.registers['a'] = initial_a
        while self.step(verbose=verbose):
            pass
        return self.registers['a']

    def run_until_out(self, verbose=False):
        "Run until the next output and return that output"

        # 1. What is the length of the current output
        current = len(self.output)

        # 2. Run until the length of the output changes
        while current == len(self.output):
            self.step(verbose=verbose)

        # 3. Return the new output
        return self.output[-1]

    def check_clock(self, verbose=False, initial_a=0):
        "Check if this initial a returns a clock signal"
        if verbose:
            print("check_clock", initial_a)

        # 1. Set up to run the machine
        self.reset()
        self.registers['a'] = initial_a
        want = NEXT[self.run_until_out()]
        count = 0

        # 2. Loop until the we get enough output
        while count < CLOCK_COUNT:

            # 3. Check that the next output is what we want
            if want != self.run_until_out():
                if verbose:
                    print("clock fail", self.output)
                return False

            # 4. Set up for the next clock signal
            want = NEXT[want]
            count += 1

        # 5. Look good to me
        if verbose:
            print("clock OK", self.output)
        return True

    def find_clock(self, verbose=False, limit=LIMIT_CLOCK):
        "Find the lowest initial a value that gives a clock signal"

        # 1. Start at the very beginning
        result = START_CLOCK

        # 2. Loop until it works
        while not self.check_clock(verbose=verbose, initial_a=result):

            # 3. Set up to try again
            result += 1

            # 4. Bail if we have been at this too long
            if result > limit > 0:
                return None

        # 5. Return the lowest value for a that generates a good clock signal
        return result

    def part_one(self, verbose=False, limit=LIMIT_CLOCK):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.find_clock(verbose=verbose, limit=limit)

    def part_two(self, verbose=False, limit=LIMIT_CLOCK):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.find_clock(verbose=verbose, limit=limit)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                    a s s e m b u n n y . p y                   end
# ======================================================================
